### PR TITLE
KT-30883 Inspection + quickfix to replace non-null assertion with mapNotNull

### PR DESCRIPTION
**Pull request by t-kameyama**
https://youtrack.jetbrains.com/issue/KT-30883

### CODE REVIEW

File: idea/resources/META-INF/plugin-common.xml [hunk 1]
```diff
@@ -3404,7 +3404,16 @@
                      level="WEAK WARNING"
                      language="kotlin"
     />
-    
+
+    <localInspection implementationClass="org.jetbrains.kotlin.idea.inspections.ReplaceNotNullAssertionWithMapNotNullInspection"
+                     displayName="Replace '!!' with 'mapNotNull'"
```
Comment (dimonchik0036) [inline idea/resources/META-INF/plugin-common.xml]: `'map' with not-null assertion should be simplified to 'mapNotNull'` looks better.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceNotNullAssertionWithMapNotNullInspection.kt [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            if (returnedExpression != postfix) return
+        } else {
+            if (blockLastStatement != postfix) return
+        }
+
+        val call = blockParent.getStrictParentOfType<KtCallExpression>() ?: return
+        if (!call.isCalling(mapNotNullFqName)) return
+
+        holder.registerProblem(
+            operationReference,
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceNotNullAssertionWithMapNotNullInspection.kt]: Should highlight `map`.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceNotNullAssertionWithMapNotNullInspection.kt [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            "Replace '!!' with 'mapNotNull'",
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceNotNullAssertionWithMapNotNullInspection.kt]: Same as above.

File: idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceNotNullAssertionWithMapNotNullInspection.kt [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import org.jetbrains.kotlin.idea.quickfix.createFromUsage.callableBuilder.getReturnTypeReference
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.typeUtil.makeNullable
+
+class ReplaceNotNullAssertionWithMapNotNullInspection : AbstractKotlinInspection() {
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/ReplaceNotNullAssertionWithMapNotNullInspection.kt]: `ConvertMapWithNotNullAssertionToMapNotNullInspection`?

File: idea/src/org/jetbrains/kotlin/idea/inspections/collections/ConvertMapWithNotNullAssertionToMapNotNullInspection.kt [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import org.jetbrains.kotlin.resolve.BindingContext
+import org.jetbrains.kotlin.resolve.lazy.BodyResolveMode
+import org.jetbrains.kotlin.types.typeUtil.makeNullable
+
+class ConvertMapWithNotNullAssertionToMapNotNullInspection : AbstractKotlinInspection() {
+    companion object {
+        private val MAP_NOT_NULL = FqName("kotlin.collections.map")
+    }
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean) = qualifiedExpressionVisitor(fun(qualifiedExpression) {
```
Comment (dimonchik0036) [inline idea/src/org/jetbrains/kotlin/idea/inspections/collections/ConvertMapWithNotNullAssertionToMapNotNullInspection.kt]: `callExpressionVisitor` for the case without a receiver
``` Kotlin
fun baz(list: List<Int?>) {
    with(list) {
        map {
            it!!
        }
    }
}
```

**Comment by goodwinnk** [comment]
@t-kameyama Thank you for the amazing work on this issue, but as Niko Knop noticed in the issue, this inspection doesn't take into account a major semantic behavior change. We are now trying to minimize amount of false positive in new inspections and considering this and changed semantic we've decided to reject the original request in the tracker.

**Pull request closed without merging**